examples/shader/extended_material_bindless.rs

```diff
@@ -0,0 +1,156 @@
+//! Demonstrates bindless `ExtendedMaterial`.
+
+use std::f32::consts::FRAC_PI_2;
+
+use bevy::{
+    color::palettes::{css::RED, tailwind::GRAY_600},
+    pbr::{ExtendedMaterial, MaterialExtension, MeshMaterial3d},
+    prelude::*,
+    render::{
+        mesh::{SphereKind, SphereMeshBuilder},
+        render_resource::{AsBindGroup, ShaderRef, ShaderType},
+    },
+    utils::default,
+};
+
+/// The path to the example material shader.
+static SHADER_ASSET_PATH: &str = "shaders/extended_material_bindless.wgsl";
+
+/// The example bindless material extension.
+///
+/// As usual for material extensions, we need to avoid conflicting with both the
+/// binding numbers and bindless indices of the [`StandardMaterial`], so we
+/// start both values at 100 and 50 respectively.
+///
+/// The `#[data(50, ExampleBindlessExtensionUniform, binding_array(101))]`
+/// attribute specifies that the plain old data
+/// [`ExampleBindlessExtensionUniform`] will be placed into an array with
+/// binding 100 and will occupy index 50 in the
+/// `ExampleBindlessExtendedMaterialIndices` structure. (See the shader for the
+/// definition of that structure.) That corresponds to the following shader
+/// declaration:
+///
+/// ```wgsl
+/// @group(2) @binding(100) var<storage> example_extended_material_indices:
+///     array<ExampleBindlessExtendedMaterialIndices>;
+/// ```
+///
+/// The `#[bindless(index_table(range(50..53), binding(100)))]` attribute
+/// specifies that this material extension should be bindless. The `range`
+/// subattribute specifies that this material extension should have its own
+/// index table covering bindings 50, 51, and 52. The `binding` subattribute
+/// specifies that the extended material index table should be bound to binding
+/// 100. This corresponds to the following shader declarations:
+///
+/// ```wgsl
+/// struct ExampleBindlessExtendedMaterialIndices {
+///     material: u32,                      // 50
+///     modulate_texture: u32,              // 51
+///     modulate_texture_sampler: u32,      // 52
+/// }
+///
+/// @group(2) @binding(100) var<storage> example_extended_material_indices:
+///     array<ExampleBindlessExtendedMaterialIndices>;
+/// ```
+///
+/// We need to use the `index_table` subattribute because the
+/// [`StandardMaterial`] bindless index table is bound to binding 0 by default.
+/// Thus we need to specify a different binding so that our extended bindless
+/// index table doesn't conflict.
+#[derive(Asset, Clone, Reflect, AsBindGroup)]
+#[data(50, ExampleBindlessExtensionUniform, binding_array(101))]
```

Comment:
The level of complexity this foists on users is pretty gnarly (both here and in shaders). Getting these indices right feels like "unnecessary implementation details".

It definitely feels like its time to start building a higher level system that abstracts over data binding and access, removing the need for developers to handle it manually. It would be very nice if "bindless" was automatic and transparent on both the rust side and the shader side (at least by default).

As a shader developer just trying to be creative (ex: I'm working on a water system in a side project at the moment), and as someone not intimately familiar with the bindless impl (which will be the majority of our shader authors), I see this complexity and think "I'll pass and eat the overhead".

Reply:
I don't like it any more than you do :) And I agree. We discussed some ideas for improvement on Discord.